File: src/load/parser-identity.ts
import definition from "../../parser-identity.json" with { type: "json" };

type ParserIdentityDefinition = {
  parserName: string;
  clarityUpstreamRelease: string;
  clarityForkRevision: string;
  exportFormatVersion: string;
};

const identity = definition as ParserIdentityDefinition;

requiredString("parserName", identity.parserName);
requiredString("clarityUpstreamRelease", identity.clarityUpstreamRelease);
requiredString("exportFormatVersion", identity.exportFormatVersion);
if (!/^[a-f0-9]{40}$/.test(identity.clarityForkRevision)) {
  throw new Error("parser-identity.json clarityForkRevision must be a full lowercase Git commit");
}

export const parserIdentity = Object.freeze({
  name: identity.parserName,
  upstreamRelease: identity.clarityUpstreamRelease,
  forkRevision: identity.clarityForkRevision,
  version: identity.clarityForkRevision,
  exporterVersion: identity.exportFormatVersion,
});

export const manifestParserIdentity = Object.freeze({
  name: parserIdentity.name,
  version: parserIdentity.version,
  upstreamRelease: parserIdentity.upstreamRelease,
  forkRevision: parserIdentity.forkRevision,
});

function requiredString(name: string, value: unknown): asserts value is string {
  if (typeof value !== "string" || value.trim() === "") {
    throw new Error(`parser-identity.json ${name} must be a non-empty string`);
  }
}

File: parser/src/main/java/lab/dota/parser/ParserIdentity.java
package lab.dota.parser;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.io.InputStream;
import java.util.Map;

record ParserIdentity(
        String parserName,
        String clarityUpstreamRelease,
        String clarityForkRevision,
        String exportFormatVersion
) {
    private static final String RESOURCE = "/parser-identity.json";

    static ParserIdentity load() {
        try (InputStream input = ParserIdentity.class.getResourceAsStream(RESOURCE)) {
            if (input == null) throw new IllegalStateException("missing " + RESOURCE);
            ParserIdentity identity = new ObjectMapper().readValue(input, ParserIdentity.class);
            identity.validate();
            return identity;
        } catch (IOException e) {
            throw new IllegalStateException("invalid " + RESOURCE, e);
        }
    }

    String parserVersion() {
        return clarityForkRevision;
    }

    Map<String, String> manifestParser() {
        return Map.of(
                "name", parserName,
                "version", parserVersion(),
                "upstreamRelease", clarityUpstreamRelease,
                "forkRevision", clarityForkRevision);
    }

    private void validate() {
        requireNonBlank("parserName", parserName);
        requireNonBlank("clarityUpstreamRelease", clarityUpstreamRelease);
        requireNonBlank("exportFormatVersion", exportFormatVersion);
        if (clarityForkRevision == null || !clarityForkRevision.matches("[a-f0-9]{40}")) {
            throw new IllegalStateException("clarityForkRevision must be a full lowercase Git commit");
        }
    }

    private static void requireNonBlank(String name, String value) {
        if (value == null || value.isBlank()) throw new IllegalStateException(name + " must be non-empty");
    }
}
